src/cargo/ops/cargo_compile/mod.rs

```diff
@@ -224,6 +224,94 @@ pub fn print<'a>(
     Ok(())
 }
 
+// Checks if a target path exists and is a source file, not a directory
+fn validate_target_path_as_source_file(
```

Comment:
Could you move this below the caller?  I've found it is generally easier to understand code when it is organized in a generally top-down fashion.